### ISSUE TITLE
JSC: Audit vxworks6 global table protections

### ISSUE THREAD

**Issue by skliper** [nasa/osal#138]
Originally part of trac #45, isolated for CCB review purposes.

Audit the usage of `semTake()` / `semGive()` calls surrounding global table accesses and adjust where needed.

**Comment by skliper** [comment]
_Imported from trac issue 115.  Created by jphickey on 2015-11-06T17:03:00, last modified: 2015-12-09T11:13:56_

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-09 11:41:06:_

Commit [changeset:ab2d6d7] is ready for CCB review, branch `trac-115-jsc-audit-table-protections`

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-09 11:54:46:_

I have reviewed the change set and it appears mostly OK however there are a few things.

* Added Comments appear to be cut-and-pasted, e.g. `QueueGetIdByName()` says it accesses the OS_task_table without locking.

* Some functions were fixed and some were not.  The same example (`OS_QueueGetIdByName()`) is still not multi-thread safe, the comment even says so.  But other ones were fixed.  Same with `OS_QueueGetInfo()` - this is still checking the "free" field before locking.  This was corrected in other "GetInfo" calls to put the "free" check inside the lock (such as `OS_TimerGetInfo()`).  

* `OS_BinSemDelete()` is another example - this deletes the semaphore while the table entry is still marked valid, allowing another thread to do a Give/Take on that semaphore even though it was deleted.

* `OS_TimerDelete()` has the opposite problem, it locks the global and frees the entry //first// and the unlocks and proceeds to access the global table again to delete the actual OS timer.  This is a race condition.


To summarize, most of the code changes are likely OK but it still leaves a lot of race conditions around.

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-16 17:17:01:_

Recommend fixing the cut-and-paste comment errors before accepting.  

It would be most desirable to fix the race conditions, but I'm OK with the code as is since there is an added comment block that warns about multi-threaded table access not being safely supported in the implementation.

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-11-17 11:02:43:_

Agree with fixing cut/paste comments.

In order to really fix all of the race conditions, a very careful look at the is needed to ensure other issues are not introduced (like undesirable timing changes, probably not such an issue in delete functions, but in read/write type functions).  Most of these remaining cases are an issue only if two different threads access and delete the same resource at the same time.  In general terms, for stand alone OSAL, this is undesirable.  But for the current common usage where OSAL is used in the context of cfe and the current app structure, this situation is not an issue (apps manage their own resources).

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-18 15:24:13:_

11/17/15 CCB: Update definition of “tmp_task_id” to be an int and add comment providing MISRA rule exemption notification/rational.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-11-20 15:37:19:_

Who is making the edit?

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-11-23 16:41:59:_

Reference commit is ready for merge to "development" branch

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-11-23 15:33:53:_

- fixed comment cut/paste error
- changed tmp_task_int from int32 to int
reference commit [changeset:e25df04]

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-11-23 23:31:49:_

Note that `OS_QueueGet` and `OS_QueuePut` still have cut-paste comment errors.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 11:52:57:_

Traceability: #89's changeset [changeset:cb5032b] has been integrated into #138 changeset [changeset:ab2d6d7].  Recommend closure of #89 in favor of #138.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 12:24:09:_

Traceability: The vxworks code changes from #92, [changeset:1a6e450] and [changeset:270c2dc] have been reflected in #138 [changeset:ab2d6d7].  Note, for background/rationale see #92 and #103 (related).  Recommend closing #92 in favor of #138.

(The above discussion is correct: there are still many locking problems throughout OSAL.  For these issues we retreated back to a more conservative stance in the face of limited time/effort.)

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 12:29:25:_

Note: similar issues were found (and not fixed) for the posix and rtems ostimer.c, #94, #95.  (From spot-checks during the work on #45 likely all the OSALs have similar problems, except for the redesign in #28?)

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 12:59:35:_

Traceability: osloader.c locking changes (OS_ModuleLoad & OS_ModuleUnload) from #99 are reflected in #138, [changeset:ab2d6d7].  Recommend closure of #99 in favor of #138 with the caveat that several osloader.c functions are (still) not thread-safe.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 15:48:04:_

Traceability: osapi.c locking changes from #103 are reflected in #138 [changeset:ab2d6d7].  Recommend closing #103 in favor of #138.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 15:55:00:_

commit: [changeset:201f00c] fixed comment-wrapping, vxworks osapi.c.

**Comment by skliper** [comment]
_Trac comment by abrown4 on 2015-11-25 16:11:15:_

Traceability: Changes from #106 are reflected in #138 [changeset:ab2d6d7].  Recommend closing #106 in favor of #138 for these improvements to be merged.

(However, osfilesys.c is one of the most thread-unsafe parts of OSAL so recommend the CCB discuss paths forward with #106 providing a reference.)

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-12-03 12:02:11:_

More information on other thread-unsafe code in the OSAL:

"looks like when a timer is deleted, the order of ops in the OS_TimerDelete function is:

   mark the slot free
   give the OS_timer_table_sem
   fetch the timer id from the table slot that just marked free
   delete the timer

once we give the semaphore with the slot marked as free it is no longer "safe" to fetch the timer id from the table slot."

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-12-09 11:07:51:_

Commits [changeset:ab2d6d7], [changeset:e25df04], [changeset:3877440], and [changeset:201f00c], included via merge [changeset:ca8b190] into merge [changeset:961b061] which is now the development branch.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-12-09 11:13:56:_

Merge of #136 #138 #139 #141 and #142 (2015-12-01) is now development.
